File: 566.java
class Solution {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        // time: O(m * n), space: O(m * n)
        int m = mat.length;
        int n = mat[0].length;
        int reshape[][] = new int[r][c];
        if (m * n == r * c) {
            for (int i = 0; i < m * n; i++) {
                reshape[i / c][i % c] = mat[i / n][i % n];
            }
        }
        else {
            reshape = mat;
        }
        return reshape;
    }
}

File: 566.py
class Solution:
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        # time: O(m * n), space: O(m * n)
        r2, c2 = len(mat), len(mat[0])
        if r2 * c2 != r * c:
            return mat
        res = []
        cur = []
        for row in range(r2):
            for col in range(c2):
                cur.append(mat[row][col])
                if len(cur) == c:
                    res.append(cur)
                    cur = []
        return res